### PR TITLE
Enable beam search

### PR DESCRIPTION
**Pull request by bocchris-aws**
Added beam search functionality with a test for bart-base model. Currently, KV cache does not work which is likely due to BART-specific preparation of decoder mask.

### CODE REVIEW

**Comment by bocchris-aws** [comment]
@michaelbenayoun Would you mind taking a look?

File: optimum/neuron/generation/utils.py [hunk 1]
```diff
@@ -218,6 +223,442 @@ def _update_past(past_key_values):
...
+        stopping_criteria: Optional[StoppingCriteriaList] = None,
+        max_length: Optional[int] = None,
+        pad_token_id: Optional[int] = None,
+        eos_token_id: Optional[Union[int, List[int]]] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        output_scores: Optional[bool] = None,
+        return_dict_in_generate: Optional[bool] = None,
+        synced_gpus: Optional[bool] = False,
+        seq_length: Optional[int] = int,
```
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py:273]: Most likely a typo here.
```suggestion
        seq_length: Optional[int] = ?,
```

File: tests/test_generate.py [hunk 2]
```diff
@@ -66,21 +68,49 @@ def _test_greedy_decoding(model_name, device="cpu", use_cache=False, decoder_onl
...
+    cpu_samples = _test_generative_decoding(model_name=model_name, device="cpu", decoder_only=decoder_only)
+
+    assert np.array_equal(cpu_samples, xla_neuron_samples_fp32), "XLA Neuron FP32 output doesn't match CPU only output"
+    assert np.array_equal(cpu_samples, xla_neuron_samples_bf16), "XLA Neuron bf16 output doesn't match CPU only output"
 
-    cpu_samples = _test_greedy_decoding(model_name=model_name, device="cpu", decoder_only=decoder_only)
+@is_trainium_test
+@pytest.mark.parametrize("model_name, use_cache, decoder_only", beam_search_testdata)
+def test_beam_search_decoding(model_name, use_cache, decoder_only):
+    os.environ['NEURON_CC_FLAGS'] = '--model-type=transformer --enable-saturate-infinity'
```
Comment (michaelbenayoun) [inline tests/test_generate.py:96]: If we need that, we should add this to the generate method no?

Comment (bocchris-aws) [reply tests/test_generate.py:96]: It is actually specific to the BART model.

**Comment by michaelbenayoun** [comment]
You can fix the styling issue by running:

```
make style
```

File: optimum/neuron/generation/utils.py [hunk 3]
```diff
@@ -549,6 +1015,44 @@ def generate(
...
+                eos_token_id=generation_config.eos_token_id,
+                output_scores=generation_config.output_scores,
+                return_dict_in_generate=generation_config.return_dict_in_generate,
+                synced_gpus=synced_gpus,
+                seq_length=input_ids_seq_length,
+                **model_kwargs,
+            )
 
         else:
             raise ValueError("Only greedy search is supported on Neuron.")
```
Comment (aashiqmuhamed) [inline optimum/neuron/generation/utils.py:1058]: Can we edit this to also include beam search @bocchris-aws?

**Review by michaelbenayoun — APPROVED** [review]

**Comment by bocchris-aws** [comment]
@michaelbenayoun is there anything on my end that I can do to make the tests pass?

**Comment by michaelbenayoun** [comment]
No, I think this is ok, you would need to be a collaborator to be able to run the CI.

**Pull request merged**